src/glb.rs

```diff
@@ -55,71 +126,134 @@ impl<'a> Glb<'a> {
+            // that of data.len().
+            .map(|json_h| data.split_at(json_h.length as usize))?;
 
-        offset += json_chunk_length as isize;
-        let bin_chunk = if data.len() as isize - offset == 0 {
-            None
+        let bin = if data.len() > 0 {
+            ChunkHeader::from_reader(&mut data)
+                .and_then(|bin_h| if let ChunkType::Bin = bin_h.ty {
+                    Ok(bin_h)
+                } else {
+                    Err(GlbError::ChunkType(bin_h.ty))
+                })
+                .and_then(|bin_h| if bin_h.length as usize <= data.len() {
+                    Ok(bin_h)
+                } else {
+                    Err(GlbError::ChunkLength {
+                        ty: bin_h.ty,
+                        length: bin_h.length,
+                        length_read: data.len(),
+                    })
+                })
+                // PANIC: we have verified that bin_h.length is no greater than
+                // that of data.len().
+                .map(|bin_h| data.split_at(bin_h.length as usize))
+                .map(|(x, _)| Some(x))?
         } else {
-            unsafe {
-                let x: *const ChunkInfo = transmute(ptr.offset(offset));
-                offset += size_of::<ChunkInfo>() as isize;
-                if (*x).length as usize > (data.len() as isize - offset) as usize {
-                    return err("BIN chunkLength exceeds length of data");
-                }
-                if &(*x).kind != b"BIN\0" {
-                    return err("BIN chunkType is not `BIN\0`");
-                }
-                let begin = offset as usize;
-                let end = begin + (*x).length as usize;
-                Some(begin..end)
-            }
+            None
         };
+        Ok((json, bin))
+    }
+}
 
-        let json = &data[json_chunk];
-        let bin = bin_chunk.map(|range| &data[range]);
-        Ok(Glb {
-            header,
-            json,
-            bin,
-        })
+impl fmt::Display for GlbError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use std::error::Error;
+        write!(f, "{}", self.description())
+    }
+}
+
+impl ::std::error::Error for GlbError {
+    fn description(&self) -> &str {
+         match *self {
+             GlbError::IoError(ref e) => e.description(),
+             GlbError::Version(_) => "unsupported version",
+             GlbError::Magic(_) => "not glTF magic",
+             GlbError::Length { .. } => "could not completely read the object",
+             GlbError::ChunkLength { ty, .. } => match ty {
+                 ChunkType::Json => "JSON chunk length exceeds that of slice",
+                 ChunkType::Bin => "BIN\\0 chunk length exceeds that of slice",
+             },
+             GlbError::ChunkType(ty) => match ty {
+                 ChunkType::Json => "was not expecting JSON chunk",
+                 ChunkType::Bin => "was not expecting BIN\\0 chunk",
```

Comment:
The `\\0` part is irrelevant to the user-facing error but it's no big deal.

Reply:
I understand, but this is made to not to confuse the user, so that user could easily look up BIN\0 chunk type in e.g. gltf specs.